crates/starknet/src/services/add_declare_transaction.rs

```diff
@@ -0,0 +1,176 @@
+use starknet_in_rust::transaction::Declare;
+use starknet_types::{
+    felt::{ClassHash, TransactionHash},
+    traits::HashProducer,
+    DevnetResult,
+};
+
+use crate::{
+    traits::StateChanger,
+    transactions::{declare_transaction::DeclareTransactionV1, StarknetTransaction, Transaction},
+    Starknet,
+};
+
+impl Starknet {
+    pub fn add_declare_transaction_v1(
+        &mut self,
+        declare_transaction: DeclareTransactionV1,
+    ) -> DevnetResult<(TransactionHash, ClassHash)> {
+        let mut declare_transaction = declare_transaction;
+
+        let class_hash = declare_transaction.contract_class.generate_hash()?;
+        declare_transaction.class_hash = Some(class_hash);
+
+        let transaction_hash = declare_transaction.generate_hash()?;
+        declare_transaction.transaction_hash = Some(transaction_hash);
+
+        let transaction = Declare {
+            class_hash: class_hash.into(),
+            sender_address: declare_transaction.sender_address.try_into()?,
+            tx_type: starknet_in_rust::definitions::transaction_type::TransactionType::Declare,
+            validate_entry_point_selector:
+                starknet_in_rust::definitions::constants::VALIDATE_DECLARE_ENTRY_POINT_SELECTOR
+                    .clone(),
+            version: declare_transaction.version.into(),
+            max_fee: declare_transaction.max_fee,
+            signature: declare_transaction.signature.iter().map(|felt| felt.into()).collect(),
+            nonce: declare_transaction.nonce.into(),
+            hash_value: transaction_hash.into(),
+            contract_class: declare_transaction.contract_class.clone().try_into()?,
+        };
+
+        transaction.verify_version()?;
```

Comment:
Why do we need this? This seems to support free transactions (v0) - do we want that? In Pythonic Devnet we had the `--allow-max-fee-zero` to allow this, but I thought we agreed we don't need this anymore.

Reply:
Changes [here](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/20/commits/c4a0300f6cabdd2738e25a0cb788b666d2112aba)